dataset/src/main/scala/frameless/functions/AggregateFunctions.scala

```diff
@@ -20,74 +20,75 @@ trait AggregateFunctions {
     new TypedAggregateAndColumn[T, Long, Long](untyped.count(column.untyped))
   }
 
-  def sum[A, T, Out](column: TypedColumn[T, A])(
+  def sum[T, Sel, Agg](column: TypedColumn[T, Sel])(
     implicit
-    summable: CatalystSummable[A, Out],
-    encoder1: TypedEncoder[A],
-    encoder2: TypedEncoder[Out])
-  : TypedAggregateAndColumn[T, A, Out] = {
+    summable: CatalystSummable[Sel, Agg],
+    encoder1: TypedEncoder[Sel],
+    encoder2: TypedEncoder[Agg])
+  : TypedAggregateAndColumn[T, Agg, Sel] = {
     val zeroExpr = Literal.create(summable.zero, encoder2.targetDataType)
     val sumExpr = expr(untyped.sum(column.untyped))
     val sumOrZero = Coalesce(Seq(sumExpr, zeroExpr))
 
-    new TypedAggregateAndColumn[T, A, Out](sumOrZero)
+    new TypedAggregateAndColumn[T, Agg, Sel](sumOrZero)
   }
 
-  def avg[A, T, Out](column: TypedColumn[T, A])(
+  def avg[T, Sel, Agg](column: TypedColumn[T, Sel])(
     implicit
-    averageable: CatalystAverageable[A, Out],
-    encoder1: TypedEncoder[A],
-    encoder2: TypedEncoder[Option[Out]]
-  ): TypedAggregateAndColumn[T, A, Option[Out]] = {
-    new TypedAggregateAndColumn[T, A, Option[Out]](untyped.avg(column.untyped))
+    averageable: CatalystAverageable[Sel, Agg],
+    encoder1: TypedEncoder[Sel],
+    encoder2: TypedEncoder[Option[Agg]]
+  ): TypedAggregateAndColumn[T, Option[Agg], Sel] = {
+    new TypedAggregateAndColumn[T, Option[Agg], Sel](untyped.avg(column.untyped))
   }
 
-  def variance[A, T](column: TypedColumn[T, A])(
+  def variance[T, Sel, Agg](column: TypedColumn[T, Sel])(
     implicit
-    variance: Variance[A],
-    encoder1: TypedEncoder[A],
-    encoder2: TypedEncoder[Option[A]]
-  ): TypedAggregateAndColumn[T, A, Option[A]] = {
-    new TypedAggregateAndColumn[T, A, Option[A]](untyped.variance(column.untyped))
+    variance: CatalystVariance[Sel, Agg],
+    encoder1: TypedEncoder[Sel],
+    encoder2: TypedEncoder[Option[Agg]]
+  ): TypedAggregateAndColumn[T, Option[Agg], Sel] = {
```

Comment:
As far as I understand there is a bit of confusion with parameter names. I would expect such signature because selecting variance of empty dataset is None. (`TypedAggregateAndColumn[T, Option[Out], Out] is TypedColumn[T, Option[Out]]`) While when we do group by it we always have value for groups (`TypedAggregateAndColumn[T, Option[Out], Out] is TypedAggregate[T, Out]`)



```scala
def variance[T, In, Out: TypedEncoder](column: TypedColumn[T, In])(
    variance: CatalystVariance[In, Out]
): TypedAggregateAndColumn[T, Option[Out], Out] = {
  import column.aencoder
  new TypedAggregateAndColumn[T, Out, Option[Out]](untyped.variance(column.untyped))
}
```